### ISSUE TITLE
Enable torch.nn.functional.softmax

### ISSUE THREAD

**Issue by michihirohorie** [torch-spyre/torch-spyre#499]
### What feature would you like to be added?

Add torch.nn.functional.softmax for spyre

### Why is this feature needed?

Torch.nn.functional.softmax is used in our target models

**Comment by michihirohorie** [comment]
Assigned to @Sunny-Anand

**Comment by Sunny-Anand** [comment]
Pytorch upstream reference: https://github.com/pytorch/pytorch/blob/e43b4d368b46eb18a8ab2ba2aab21e3d581cd74e/torch/_decomp/decompositions.py#L1297 
Amax and Sum is not supported yet for FP32 to allow to use the upstream softmax.  
Currently amax and sum are nort supported: https://github.com/torch-spyre/torch-spyre/blob/main/torch_spyre/_inductor/constants.py#L51-L73.
Previous issue for amax enablement closed: https://github.com/torch-spyre/torch-spyre/issues/583 with tests using fp16.
Amax FP16 pass results pls share @CaiwuChen to confirm

Sum fp16 is passing , FP32 is blocked https://github.com/torch-spyre/torch-spyre/issues/1871 
Backend seems to enable FP32 sum https://github.ibm.com/ai-chip-toolchain/deeptools/pull/4287

**Comment by CaiwuChen** [comment]
Torch.sum():

<img width="892" height="578" alt="Image" src="https://github.com/user-attachments/assets/b8572d9b-e630-4204-b498-b815479c8c9b" />
FP16 tests are passing, FP32 test are failing.


Torch.amax():

<img width="1645" height="1187" alt="Image" src="https://github.com/user-attachments/assets/64c9b4d3-765d-49c0-b6a6-5f59db2b81b9" />
All existing test pass but the default dtype is `torch.float16` as of commit f2dfd86

https://github.com/torch-spyre/torch-spyre/blob/f2dfd86/tests/inductor/utils_inductor.py#L45

**Comment by moriohara** [comment]
@CaiwuChen There are a few sub issues to be solved to fully enable FP32 Softmax.
- FP32 reduction (masking or padding necessary when the stick dimension size is not a multiple of the stick size. This requires a backend support.
- Non-Standard ElementArrangement where a data conversion to change the element size (i.e. fp16 to/from fp32) changes the arrangement of the elements on the stick dimension, which https://github.com/torch-spyre/torch-spyre/issues/2508 and https://github.com/torch-spyre/torch-spyre/issues/2583 are also dealing with.

**Comment by CaiwuChen** [comment]
Symptom: FP32 softmax fails on all reduce-on-last-dim cases (fp32_2d_dim1, fp32_3d_dim2, fp32_4d_dim3) with DtException: [Transformation: 4Bsplat] at backend compile time, or a 280s hardware hang timeout when the crash is suppressed.

Root Cause: When softmax operates on the last dim, the amax and sum ops it decomposes into both reduce along the stick dim. The backend DDL template (summeanmaxexx2_fp32.ddl) requires a SPLAT + REDUCE sequence on the SFP unit to fold the 32 stick elements into a scalar. The DDC transformation transformFor4BsplatRead() has a bug when the FIFO result is consumed by an opaque REDUCE instruction — it crashes at compile time. Bypassing the crash still results in a hardware hang, confirming the entire FP32 stick-reduce path is currently broken in the backend. FP32 softmax on non-last dims is unaffected, as those cases use the sumnonstick/maxnonstick path which works correctly.

Why the Frontend Cannot Work Around It: The natural workaround would be to assign a nonstick layout to the graph input (stick dim on a non-last dim), routing amax/sum through the working nonstick path without touching the broken stick path. However, this requires ReStickifyOpHBM to convert the graph input from its physical stick layout to a nonstick layout at runtime — and ReStickifyOpHBM does not yet support FP32 (frontend issue #2544, blocked by backend issue #2560).

Required Fix: Implement FP32 support in ReStickifyOpHBM.



## Test
```
import torch

torch.manual_seed(42)

def test(x):
    return torch.sum(x, dim=-1)/torch.amax(x, dim=1)

x = torch.randn(512, 1024, dtype=torch.float32)

compiled = torch.compile(test, backend="inductor")
output = compiled(x.to("spyre")).cpu()
expected = test(x)

torch.testing.assert_close(output, expected, atol=1e-1, rtol=1e-1)
```

Above Test give me timeout error.
```
(dt-inductor) [caiwu@caiwu-spyre-dev-pf torch-spyre]$ python test_amax.py
 ERRR 07.07.2026 21:01:43.737197 [           response_worker.cpp: 441]   Timed-out: PendingRequest(Node name: sdsc_fused_amax_0_fef8l9lz/spyreCodeDir/bundle.mlir#2, tag: 4, PipelineId(COMPUTE), QueuingMode(REGULAR), batch_id: 4, cb_idx: 0, state: PendingRequestState(TIMED_OUT), issued 280000 ms ago)
 ERRR 07.07.2026 21:01:43.737402 [                  ras_base.hpp:  74] {"action":"information","category":"hardware","code":"0x7b1b","description":"A hardware error was detected in the compute control block response.","message":"Compute CB hardware error detected","name":"RAS::RUNTIMESCHEDULER::ComputeHardwareError","severity":"ERROR","step":"Open ticket","type":"runtime_error"}
 ERRR 07.07.2026 21:01:43.738785 [           response_worker.cpp: 912] No ResponseBlock received within timeout of 280 seconds, num_pending_requests: 2, seconds waited: 280
 ERRR 07.07.2026 21:01:43.738813 [                  ras_base.hpp:  74] {"action":"information","category":"software","code":"0x1acc","description":"An expected response block was not received.","message":"Fail on RB time-out","name":"RAS::RESPONSEWORKER::RbTimeOut","severity":"ERROR","step":"Open ticket","type":"runtime_error"}
terminate called after throwing an instance of 'std::runtime_error'
  what():  {"action":"information","category":"software","code":"0x1acc","description":"An expected response block was not received.","message":"Fail on RB time-out","name":"RAS::RESPONSEWORKER::RbTimeOut","severity":"ERROR","step":"Open ticket","type":"runtime_error"}
Signal Received: 6 (Aborted)
Signal Received from pid=121964 
```

**Comment by CaiwuChen** [comment]
Symptom: h2f softmax fails on all dims with InductorError at frontend compile time.

Root Cause: dl16tofp32(x) produces buf0 with EA=DL16_TO_FP32 (staggered layout, stagger dim = last dim, size=64). amax(buf0, dim=d) reduces on the softmax dim, not the stagger dim, so buf1 is STANDARD with stagger-dim size=64. sub(buf0, buf1) mixes DL16_TO_FP32 + STANDARD where the STANDARD side has stagger-dim size=64 — violating the broadcast rule — and compilation aborts.

Why PR #2927 (RMSNorm fix) Does Not Help: PR #2927 allows mixed EA only when the STANDARD operand has stagger-dim size=1. RMSNorm always reduces on the last dim (= stagger dim), so this holds. Softmax reduces on an arbitrary dim; the stagger dim survives at size=64 for dim≠-1, so the rule never applies.


Required Fix: Non-last dims require either hardware support for consuming DL16_TO_FP32 in standard FP32 ops (#2205), or an explicit restickify after dl16tofp32 to normalize to STANDARD before the softmax chain.

**Comment by moriohara** [comment]
@CaiwuChen Thank you so much for looking into this. There are a few tricks we need to play with the data type conversion affecting the element size:
- a data conversion op (fp16 to fp32 and vice versa) changes the element arrangement at the stick dimension, where elements in the dimension has a non-standard order in the dimension.
- a reduction op (e.g. sum) at the stick dimension reduces those elements into a single element in a stick (i.e. sparse), which is essentially the same as the standard arrangement.
- a reduction op (e.g. sum) at a non-stick dimension reduces elements in the dimension into a single element. The element arrangement stays the same, which can be restored to the standard by converting the data type from fp32 to fp16.
PR #2927 has been merged. You can look into changes from the PR in propagate_layouts.py to see how the element arrangement is handled. There may be some cases where the current logic is not robust enough.
You can also post a repro code for further discussions.
